src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -557,10 +557,25 @@ QSizeF QgsSymbolLegendNode::drawSymbol( const QgsLegendSettings &settings, ItemC
   double widthOffset = 0;
   double heightOffset = 0;
 
+  double maxSymbolSize = settings.maxSymbolSize();
+  double minSymbolSize = settings.minSymbolSize();
+
+  std::unique_ptr<QgsMarkerSymbol> minMaxSizeMarkerSymbol;
   if ( QgsMarkerSymbol *markerSymbol = dynamic_cast<QgsMarkerSymbol *>( s ) )
   {
     // allow marker symbol to occupy bigger area if necessary
     double size = markerSymbol->size( *context ) / context->scaleFactor();
+
+    //If marker size exceeds max marker size, clone symbol and set max size in mm
+    if ( ( maxSymbolSize > 0 && size > maxSymbolSize ) || ( minSymbolSize > 0 && size < minSymbolSize ) )
+    {
+      minMaxSizeMarkerSymbol.reset( dynamic_cast<QgsMarkerSymbol *>( s->clone() ) );
+      minMaxSizeMarkerSymbol->setSize( size > maxSymbolSize ? maxSymbolSize : minSymbolSize );
+      minMaxSizeMarkerSymbol->setSizeUnit( QgsUnitTypes::RenderMillimeters );
+      s = minMaxSizeMarkerSymbol.get();
+      size = ( size > maxSymbolSize ) ? maxSymbolSize : minSymbolSize;
+    }
```

Comment:
Could you move this out to a function in QgsSymbolLayerUtils? (same with line symbol scaling)